src/compiler/checker.ts

```diff
@@ -19381,7 +19381,24 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
         for (let i = 0; i < paramCount; i++) {
             const sourceType = i === restIndex ? getRestTypeAtPosition(source, i) : tryGetTypeAtPosition(source, i);
-            const targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            let targetType = i === restIndex ? getRestTypeAtPosition(target, i) : tryGetTypeAtPosition(target, i);
+            if (i === restIndex && targetType && sourceType && isTupleType(sourceType)) {

```

Comment:
You're intentionally not fixing the case where `sourceType` is a union, right?

Reply:
> You're intentionally not fixing the case where sourceType is a union, right? I think that's ok because I think the code would become a lot more complicated, but I wanted to ask.

Yeeah. It's likely that this could be fixed by wrapping the body of this loop, executing it for each union member, and checking if any of the iterations return ~ `Ternary.False`. Maybe that could be explored in a follow-up PR - but for the time being, I think that it's best to improve this incrementally. It's way easier to follow this diff here than it would be to follow it if I would try to handle union sources as well.

> Same for when targetType is an array I guess.

Hmm, I think this is actually already covered - it's just that for this case we don't need to do anything special, so we forward the type so it can be checked by the latter logic as-is. Maybe I'm missing something though - do you have any particular test case in mind? The main purpose behind the added code is to "adjust" the target tuple types - their lengths etc, as without that we simply fail on required elements checks etc (but when comparing signatures we can actually often ignore extra positions that are not utilized by the source signature).